Try: make logic for conditional editor_styles consistent

mikachan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
In https://github.com/WordPress/gutenberg/pull/44640, we added a check for `wp-block-styles` theme support before including the `wp-block-library-theme` CSS in the editor.

This seemed to work well in most cases, however, there are still some cases where this CSS is included. As seen in this issue for TT3 https://github.com/WordPress/twentytwentythree/issues/295, the opinionated width of 100px for the separator is still being loaded from the theme.css file in the editor, but this isn't used on the front end.

| Editor | Front end |
| ----- | ---------- |
| <img width="398" alt="image" src="https://user-images.githubusercontent.com/1645628/198323197-f643c79b-e18d-4b06-92e3-74157cd77f19.png"> | <img width="485" alt="image" src="https://user-images.githubusercontent.com/1645628/198323383-ea7aaaca-767e-443a-9be9-bd0ee9cb8102.png"> |

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The editor styles should match the front end.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
This PR is an attempt to never allow the `wp-block-library-theme` CSS to be loaded under the conditions we introduced in https://github.com/WordPress/gutenberg/pull/44640.

I've made the logic consistent between /template-parts-screen.php and /script-loader.php, so the condition to include the theme.css file is only if the active theme supports `wp-block-styles` AND there are no `editor_styles` set.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Activate TT3 / a theme that doesn't support `wp-block-styles` AND doesn't include `add_editor_style( 'style.css' );`
2. Insert a Separator block, see the difference in width between the editor and front end

Example separator markup for copy/pasting:

```
<!-- wp:separator -->
<hr class="wp-block-separator has-alpha-channel-opacity"/>
<!-- /wp:separator -->
```

mikachan committed:
Update logic to match client-assets

mikachan committed:
Move gutenberg_override_style into if statement

mikachan commented on lib/compat/wordpress-6.1/template-parts-screen.php:
```diff
@@ -187,7 +187,7 @@ static function( $classes ) {
 
 	if (
-		current_theme_supports( 'wp-block-styles' ) ||
+		current_theme_supports( 'wp-block-styles' ) &&
```
This isn't directly related to the issue this PR is trying to address, but I noticed this logic wasn't quite the same as what is used in `client-assets.php`.

mikachan replied:
This needs to be updated in Core in [wp-admin/site-editor.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/site-editor.php#L148).

scruffian commented:
I'm not sure if something changed in TT3 or in Gutenberg, but this isn't working for me, sorry :(

mikachan committed:
Merge branch 'trunk' into fix/editor-block-styles

mikachan committed:
Update logic to match client-assets

mikachan committed:
Merge branch 'trunk' into fix/editor-block-styles

# Conflicts:
#	lib/client-assets.php

mikachan committed:
Move wp-block-library-theme inside current_theme_supports logic

mikachan committed:
Revert "Move wp-block-library-theme inside current_theme_supports logic"

This reverts commit 89c3d7d2e3a8e63fd401cf269e3e2196e760dfcf.

mikachan committed:
Remove $editor_styles check

mikachan commented:
> I'm not sure if something changed in TT3 or in Gutenberg, but this isn't working for me, sorry :(

Thanks for testing this out. I checked this just after you left this comment, and I was seeing the same as you where the fix here didn't have any effect. I tried debugging different things but I didn't get anywhere, so I left it. Now however it does seem to be working again. I'm not sure what's different! But there were several hundred commits between this and trunk, so I've rebased 😅

I think I'm understanding the problem more now, as this seems very specific to TT3. TT3 doesn't have a functions.php, which means it doesn't provide any editor styles. I believe this means this logic, `( ! is_array( $editor_styles ) || count( $editor_styles ) === 0 )`, is causing the theme.css to load in the editor.

~Is this still the logic that we want to determine including the theme.css? I've updated this PR to remove it, which fixes the issue I'm seeing with the separator and means theme.css isn't included.~

~If we still need this check, I think the order needs to be switched so $editor_styles are checked before the theme supports, like this: `( ! is_array( $editor_styles ) || count( $editor_styles ) === 0 ) && current_theme_supports( 'wp-block-styles' )`.~

Nevermind, I don't think this is working either. It's certainly related to the missing `add_editor_style( 'style.css' );` in a functions.php file for TT3, but the above changes I described don't seem to make a difference. If I remove all the calls to the theme.css file in GB, wp-admin/load-styles.php still seems to load these styles in regardless.

mikachan committed:
Revert "Remove $editor_styles check"

This reverts commit b3b368bc57c62158d271f4c3c602b3e2e4c1de49.

mikachan commented:
I've narrowed this down to the logic for including the theme.css stylesheet not being consistent, namely in script-loader.php. I've updated the logic in this PR to what I think is correct.

Also, I can confirm how to replicate this now. The theme should not support `wp-block-styles` AND not include `add_editor_style( 'style.css' );` / a functions.php file.

However, the only way I could successfully test this was to update the script-loader.php file in Core. I'd appreciate if anyone could confirm that this should be an update to Core rather than in Gutenberg.

I think this update is important to allow themes to not include a functions.php file if it's not necessary (like in Twenty Twenty-Three).

mikachan committed:
Merge branch 'trunk' into fix/editor-block-styles

mikachan committed:
Remove empty line

carolinan commented:
I am not sure what may have changed but I don't see the incorrect editor styles being loaded if I activate Gutenberg **trunk**. ?

mikachan commented:
Thanks for taking a look, @carolinan. 

I'm still seeing the incorrect width on the separator block on trunk in the editor:

| Editor on trunk | Editor with this PR | Front end |
| ------ | ------ | --------- | 
| <img width="1085" alt="image" src="https://user-images.githubusercontent.com/1645628/214803022-5a597432-6d29-41c6-b6cd-2256e2ddd982.png"> | <img width="1085" alt="image" src="https://user-images.githubusercontent.com/1645628/214804025-065aff4b-0945-4a62-b0c1-87b77b22ed26.png"> | <img width="1085" alt="image" src="https://user-images.githubusercontent.com/1645628/214803724-29777f8f-87dc-4172-b266-1833d3879894.png"> |

I have to apply the changes in this PR directly to Core (my local version of wordpress-develop) to make this work, rather than just Gutenberg. I'm guessing that's because I'm not sure how the compat files work!

carolinan commented:
Hm.. I was adding a new separator in the block editor, there it looks fine.
But when I look at the one already saved in the template in the site editor, it is too short.

carolinan commented:
> I'm not sure how the compat files work!

We need someone to do a zoom call with us and explain it :)

mikachan commented on lib/compat/wordpress-6.2/script-loader.php:
```diff
@@ -80,7 +80,10 @@ function gutenberg_resolve_assets_override() {
 	);
 
-	if ( current_theme_supports( 'wp-block-styles' ) ) {
+	if (
+		current_theme_supports( 'wp-block-styles' ) &&
+		( ! is_array( $editor_styles ) || count( $editor_styles ) === 0 )
+	) {
```
This needs to be updated in Core in [wp-includes/script-loader.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/script-loader.php#L1627).

mikachan commented on lib/compat/wordpress-6.1/template-parts-screen.php:
```diff
@@ -178,7 +178,7 @@ static function( $classes ) {
 
 	if (
-		current_theme_supports( 'wp-block-styles' ) ||
+		current_theme_supports( 'wp-block-styles' ) &&
```
This needs to be updated in Core in [wp-admin/site-editor.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/site-editor.php#L148).

jffng commented:
As a sanity check, this PR is working for me in the front-end, post editor, and template **part** editor!

Unless you want to write some compatibility code to overwrite what assets are enqueued in the site editor, I do think the easiest fix is to patch this in core as you highlighted.

My impression is that the compat files all work a little differently depending on the context, and the context is always changing... I would definitely join a call with an overview though of the major places where these compat files are necessary and how to approach making changes there.

mikachan commented:
Thanks @jffng! For completeness (and hopefully to make things easier), I've opened a core patch as well: https://github.com/WordPress/wordpress-develop/pull/3916

carolinan commented:
I ran into extra padding being added to group blocks that has a background, only in the site editor.
Theme, Twenty Twenty-Three:
I can confirm that the core patch solves this problem.
Before:
![Screenshot 2023-02-02 at 06 20 48](https://user-images.githubusercontent.com/7422055/216239291-a4ee2384-e64c-43d0-98c9-1bfbfbff0243.png)

After, the themes global padding is applied correctly:
![Screenshot 2023-02-02 at 06 34 57](https://user-images.githubusercontent.com/7422055/216240497-d8e0da75-c6d2-400a-a320-2a663a434b3a.png)

carolinan commented:
I do not know if this needs to be merged both into Gutenberg and core? If so it needs to be refreshed and moved to the  6.2 folder.

I know that if we want it to be in 6.2 we need to add the backport label..

mikachan commented:
> I do not know if this needs to be merged both into Gutenberg and core? If so it needs to be refreshed and moved to the 6.2 folder.

Me neither! @ndiego, is this something you could help with? To summarise, this PR is changing files in the lib/compat files, but there is also a [PR open for core](https://github.com/WordPress/wordpress-develop/pull/3916/) to make the same changes there.

ndiego commented:
> I do not know if this needs to be merged both into Gutenberg and core? If so it needs to be refreshed and moved to the 6.2 folder.
> 
> Me neither! @ndiego, is this something you could help with? To summarise, this PR is changing files in the lib/compat files, but there is also a https://github.com/WordPress/wordpress-develop/pull/3916/ to make the same changes there.

cc @Mamaduka @ntsekouras are you able to provide guidance on this? I am a little unsure myself. Thanks! 🙏

glendaviesnz commented:
@mikachan my understanding is that if this is to go into 6.2 you would need to:

- Add a new file `lib/compat/wordpress-6.2/template-parts-screen.php`
- Copy your modified `gutenberg_template_parts_screen_init` method into this file and rename it to `gutenberg_template_parts_screen_init_6_2`
- Also add an action to this new file `add_action( 'admin_enqueue_scripts', 'gutenberg_template_parts_screen_init_6_2', 5 );`  - not 100% sure what priority to use here, I think 5 will work as no priority is set on the 6.1 version.
-  Add the new file to the 6.2 section of `lib/load.php`
 - Revert the change that you made in `lib/compat/wordpress-6.1/template-parts-screen.php` 
 - Add the `Backport to WP RC/Beta` label
 - Add details to the relevant sections of https://github.com/WordPress/gutenberg/issues/47187

Sing out if it does get through to the backporting stage. I am happy to help out if you are not familiar with the process.

Mamaduka commented:
In this case, I think it's okay to modify `gutenberg_template_parts_screen_init` directly.

## Why?

* The custom screen is only available when the Gutenberg plugin is active [with WP 6.0](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.1/template-parts-screen.php#L40).
* We'll drop WP 6.0 support a few weeks after 6.2 is released, and all code living in `compat/6.1` will be removed.

mikachan commented:
Thanks, both!

> In this case, I think it's okay to modify gutenberg_template_parts_screen_init directly.

Do you mean directly in Core, like in https://github.com/WordPress/wordpress-develop/pull/3916?

Mamaduka commented:
Sorry for the late reply, @mikachan!

Re `template-parts-screen.php`: The changes are good to merge as they're 👍

mikachan commented:
Thanks, @Mamaduka!

So is it OK to merge this PR as it is, or do I need to follow the backporting steps outlined above by @glendaviesnz?

carolinan committed:
Merge branch 'trunk' into fix/editor-block-styles

mikachan commented:
Closing this, as @hellofromtonya kindly tested and merged the Core patch at https://github.com/WordPress/wordpress-develop/pull/3916.

Thanks for everyone's help with this one! 🙇‍♂️
